src/_tests/test_3-train.py

```diff
@@ -0,0 +1,251 @@
+import os
+import time
+
+import numpy as np
+import pytest
+import torch
+
+from network import build_network
+
+from ..train import build_criterion, prepare_data_and_mask, train_network
+from ..utils import construct_dataloaders, normalization, qlty_crop
+
+
+def test_load_yaml(parameters_dict):
+    assert isinstance(parameters_dict, dict)
+    assert "io_parameters" in parameters_dict
+    assert "model_parameters" in parameters_dict
+
+
+def test_io_parameter_validation(io_parameters):
+    assert io_parameters.data_tiled_uri == "http://data/tiled/uri"
+    assert io_parameters.data_tiled_api_key == "a1b2c3"
+    assert io_parameters.mask_tiled_uri == "http://mask/tiled/uri"
+    assert io_parameters.mask_tiled_api_key == "d4e5f6"
+    assert io_parameters.seg_tiled_uri == "http://seg/tiled/uri"
+    assert io_parameters.seg_tiled_api_key == "g7h8i9"
+    assert io_parameters.uid_save == "pytest1"
+    assert io_parameters.uid_retrieve == "pytest1"
+    assert io_parameters.models_dir == "."
+
+
+def test_network_name(network_name):
+    assert network_name == "DLSIA TUNet"
+
+
+def test_model_parameter_validation(model_parameters):
+    assert model_parameters.network == "DLSIA TUNet"
+    assert model_parameters.num_classes == 2
+    assert model_parameters.num_epochs == 2
+    assert model_parameters.optimizer == "Adam"
+    assert model_parameters.criterion == "CrossEntropyLoss"
+    assert model_parameters.weights == "[1.0, 0.5]"
+    assert model_parameters.learning_rate == 0.1
+    assert model_parameters.activation == "ReLU"
+    assert model_parameters.normalization == "BatchNorm2d"
+    assert model_parameters.convolution == "Conv2d"
+    assert model_parameters.qlty_window == 4
+    assert model_parameters.qlty_step == 2
+    assert model_parameters.qlty_border == 1
+    assert model_parameters.shuffle_train is True
+    assert model_parameters.batch_size_train == 1
+    assert model_parameters.batch_size_val == 1
+    assert model_parameters.batch_size_inference == 2
+    assert model_parameters.val_pct == 0.2
+    assert model_parameters.depth == 2
+    assert model_parameters.base_channels == 4
+    assert model_parameters.growth_rate == 2
+    assert model_parameters.hidden_rate == 1
+
+
+@pytest.fixture
+def raw_data(tiled_dataset):
+    data, _ = prepare_data_and_mask(tiled_dataset)
+    yield data
+
+
+@pytest.fixture
+def mask_array(tiled_dataset):
+    _, mask = prepare_data_and_mask(tiled_dataset)
+    yield mask
+
+
+def test_data_and_mask(raw_data, mask_array):
+    assert raw_data.shape == (2, 6, 6)
+    assert mask_array.shape == (2, 6, 6)
+
+
+@pytest.fixture
+def normed_data(raw_data):
+    normed_data = normalization(raw_data)
+    yield normed_data
+
+
+def test_normalization(normed_data):
+    assert type(normed_data) is np.ndarray
+    assert np.min(normed_data) == 0
+    assert np.max(normed_data) == 1
+
+
+@pytest.fixture
+def data_tensor(normed_data):
```

Comment:
If this is only being used in one test, it not not need to be a fixture. Just call the one line in your test

```
 data_tensor = torch.from_numpy(normed_data)

```

Reply:
Good catch, modified in the next commit